arch/eventloop_posix_tcp.c

```diff
@@ -167,15 +167,17 @@ TCP_listenSocketCallback(UA_ConnectionManager *cm, UA_FD fd,
 
     /* The socket has opened. Signal it to the application. The callback can
      * switch out the context. So put it into a temp variable.  */
-    void *ctx = cm->initialConnectionContext;
-    cm->connectionCallback(cm, (uintptr_t)fd, &ctx, UA_STATUSCODE_GOOD,
+    void *fdctx_address = NULL;
+    UA_EventLoop_getFDContext(cm->eventSource.eventLoop, fd, &fdctx_address);
+    cm->connectionCallback(cm, (uintptr_t)fd, (void **) fdctx_address, UA_STATUSCODE_GOOD,
```

Comment:
I don't get the change.
This is about initializing the context of the new connection.
Why do you want to forward the context of the parent fd?
The pointer to the cm and the initialContext should be enough.

At any rate, you want to forward a "true" double-pointer.
So that `connectionCallback` can switch out the pointer to something else if it wants to.

Reply:
From my point of view, within the TCP_listenSocketCallback, the listensocket will be called with its registered context which can be changed. But this extra work with **getFDContext** done here is unnecessary because the correct address is already passed via _fdcontext_ to this same function.

I modified the code so the existing address is used instead.

Reply:
Of course its a different situation when we never want to actually call the `cm->connectionCallback` on the listenSocket with its context. 

Then we maybe should change `cm->connectionCallback` to use `newsockfd` and not `fd` shouldn't we?

Reply:
So there are actually two versions that work and one version that doesn't with the current implementation of the server integration:
Within `TCP_listenSocketCallback:

1. Mixing (Doesn't work): calling `cm->connectionCallback` with the parent `fd` but the soon to be child `ctx` of `newsockfd`
2. Child only (Does work): calling `cm->connectionCallback` with the `newsockfd` and its soon to be child context `ctx`
3. Parent only (Does work): calling `cm->connectionCallback` with the parent `fd` and its already registered `fdcontext` 

Why "mixing" does not work and the others do:

- For the server integration I assume each connection id (e.g. `fd` or `newsockfd` in `TCP_listenSocketCallback`) will always be called with its registered `fdcontext`.
- So if the passed context is the initial one, I will store the connection id within the context so that it can be used during calls of the connection manager.
- If I primarily call the initial context with another connectionId and subsequently call it with a new id it will fail, because the wrong connectionId is registered in the connection context. (i.e. before my change `TCP_listenSocketCallback` would call `cm->connectionCallback` with {`parent-fd`, `child-ctx`} which stores `parent-fd` for any calls to the connection manager and subsequently `TCP_connectionSocketCallback` would call `cm->connectionCallback` with {`child-fd`, `child-ctx`})

Reply:
If for any reason, we need those calls that use the parent `fd` with the child `ctx` (**Mixing**) the server integration callback can be changed, so that it modifies the `connectionId` of the passed context on every call.

Reply:
>  I assume each connection id (e.g. fd or newsockfd in TCP_listenSocketCallback) will always be called with its registered fdcontext

You are right. It should be that way.
I prefer option 2 (child only).
If required, the parent fd and context can always be passed on via the initialContext.
And that can be solved in the individual implementation "outside" the main eventloop code.

What do you think?

Reply:
Using the connected socket (the "child" of the listen-socket) and its child-context sounds good to me. After all the callback is named `connectionCallback` which might indicate it will not be called for the initial listen-socket but for any socket that is connected.

Do you think there could be a situation where we want the `cm->connectionCallback` to get the listen-socket and its context for any reason? Because like this the registered context of the listenSocket will never be used anyhwere, is that correct?

Also, what do you mean by 
> If required, the parent fd and context can always be passed on via the initialContext.

?

Reply:
> Because like this the registered context of the listenSocket will never be used anyhwere, is that correct? 

Yes. I think that is the case.
At any rate, we forward the connectionManager pointer. That ought to be enough.
And if not, the initialContext can be set up to forward additional information when a new socket goes into the userland callback for the first time.

That also answers the question below.